working/0546-patterns/patterns-feature-specification.md

```diff
@@ -2159,17 +2209,233 @@ To match a pattern `p` against a value `v`:
+  late final $match_1_lt4 = $match_1 < 4;
+  late final $match_0_isEven = $match_1.isEven;
+  late final $match_0_isEven_eqtrue = $match_0_isEven == true;
+
+  if ($match_length_eq2 &&
+      $match_0_eq1 &&
+      $match_length_eq2 &&
+      $match_1_lt4) {
+    print('first');
+  } else if ($match_length_eq2 &&
+      $match_0_isEven_eqtrue) {
+    var a = $match_1;
+    print('second $a');
+  }
+}
+```
+
+Note that every method call is encapsulated in a `late` variable ensuring that
+it only gets invoked once even when used by multiple patterns.
+
+It works like this:
+
+1.  At compile time, after type checking has completed, we associate an
+    *invocation key* with every member call or record field access potentially
+    made by each pattern.
+
+2.  At runtime, whenever the runtime semantics say to call a member or access a
+    record field, if a previous call or access with that same invocation key has
+    already been evaluated, we reuse the result.
+
+3.  Otherwise, we invoke the member or field access now and associate the result
+    with that invocation key for future calls.
+
+Let an *invocation key* comprise:
+
+*   A possibly absent parent invocation key.
+*   A possibly absent extension type. If the invocation represents an extension
+    member call, this tracks the type the call was resolved to.
+*   A member name.
+*   A possibly empty list of argument constant values.
+
+Two invocation keys are equivalent if and only if all of these are true:
+
+*   They both have parent invocation keys and the keys are equivalent or
+    neither of them have parent invocation keys.
+*   The extension types refer to the same type or are both absent.
+*   The member names are equivalent.
+*   The argument lists have the same length and all corresponding pairs of
+    argument constant values are identical.
+
+*In other words, they're equal if all of their fields are equal in the obvious
+ways.*
+
+The notation `parent : (name, args)` creates an invocation key with parent
+`parent`, no extension type `type`, member name `name`, and argument list
+`args`. The notation `parent : type(name, args)` creates an invocation key with
+parent `parent`, extension type `type`, member name `name`, and argument list
+`args`.
+
+Given a set of patterns `s` matching a value expression `v`, we bind an
+invocation key to each member invocation and record field access in `s` like so:
+
+1.  Let `i` be an invocation key with no parent, no extension type, named
+    `this`, with an empty argument list. *This is the root node of the
+    invocation key tree and represents the matched value itself.*
+
+2.  For each pattern `p` in `s` with parent invocation `i`, bind invocation keys
+    to it and its subpatterns using the following procedure:
+
+To bind invocation keys in a pattern `p` using parent invocation `i`:
+
+*   **Logical-or** or **logical-and**:
+
+    1.  Bind invocations in the left and right subpatterns using target `i`.
```

Comment:
You switch to using "target" here instead of "parent invocation".  I think (but could be wrong), you mean the same thing?  If not, probably explain?  If so, maybe just use one term?

Reply:
Oops, yes. I went back and forth on the terminology and left it inconsistent. Fixed.